### ISSUE TITLE
[no-unsafe-assignment] Doesn't catch unsafe readonly number index assigment

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#81]
It currently catches this:
```ts
        type MutableA = {
          [P in string]: MutableA;
        };
        type ReadonlyA = {
          readonly [P in string]: ReadonlyA;
        };
        const readonlyA: ReadonlyA = {};
        const mutableA: MutableA = readonlyA; // flagged
```

but not this:

```ts
        type MutableA = {
          [P in number]: MutableA;
        };
        type ReadonlyA = {
          readonly [P in number]: ReadonlyA;
        };
        const readonlyA: ReadonlyA = {};
        const mutableA: MutableA = readonlyA; // not flagged; should be
```
